### DISCUSSION TITLE
Support for Next-Auth

### DISCUSSION THREAD

**Discussion by VinayaSathyanarayana** [keystonejs/keystone#5347] [Feature requests]
[Next-Auth](https://next-auth.js.org/) is a fairly mature framework. KeystoneJs should support Next Auth natively for Authentication

**Comment by borisno2** [comment]
A couple of us had a little chat on [Slack](https://keystonejs.slack.com/archives/C01STDMEW3S/p1617434323051300). I have had a little go of getting this working - https://github.com/OpenSaasAU/keystone/tree/nextjs-auth/packages-next/nextjs-auth. I am hoping there is a better way to do this.
It is based on `@keystone-next/auth` with a couple of changes, so much clean up is required - more of a POC/playground at the moment.
Currently the NextAuth session is setup and then on a `/signin` page a graphql where I have then changed
```javascript
resolvers: {
      Mutation: {
        async [gqlNames.authenticateItemWithPassword]
```
To `const nextSession = await getSession({ req });` from NextAuth and check for a valid User using the `sub` with the subject being the idenitity field setup in Keystone. Based on this start a Keystone Session in the Graphql Mutation.
```javascript
const sessionToken = await context.startSession({ listKey, itemId: result.item.id  });
return { sessionToken, item: result.item };
```
I feel it would be much better to handle all this in the `callbacks` in https://github.com/OpenSaasAU/keystone/blob/nextjs-auth/packages-next/nextjs-auth/src/pages/NextAuthPage.tsx if possible, I wasn't sure if it is possible to get the `KeystoneContext` at this point in order to `context.startSession({})`

This file is copied to  `pages/api/auth/[...nextauth.js]`
```javascript
let filesToWrite: AdminFileToWrite[] = [
      {
        mode: 'write',
        outputPath: 'pages/api/auth/[...nextauth].js',
        src: authTemplate({ gqlNames, identityField, secretField, }),
      },
```

Happy for any input/direction...

**Reply by VinayaSathyanarayana** [reply]
@JedWatson Your attention is requested

**Reply by borisno2** [reply]
Hey, not sure if anyone has looked at this and has any feedback. I did notice that with the latest version on `@keystone-next/auth` the session info is brought into the auth module. Wondering if a better option is to use the `SessionStrategy` and override the `get` to first check for a valid next-auth token

Just wanting to get social auth/oidc working (specifically auth0) as I have a potential use case for Keystone, but auth0 is a requirement, if this is better solved another way (ie not with next-auth), that would be awesome too.

Cheers!

**Reply by VinayaSathyanarayana** [reply]
Any update on this from the core group?

**Comment by raveling** [comment]
Thanks @VinayaSathyanarayana and @borisno2 🙌 

We acknowledge that social login could be better, and appreciate the suggestion of Next-Auth.

This isn't part of our plans for Keystone 6 `General Availability` release, or the immediate body of work that comes after it. So likely something for 2022.

Hope this helps!

**Reply by borisno2** [reply]
No worries @raveling! All good now! https://github.com/OpenSaasAU/keystone-nextjs-auth 👍

**Reply by VinayaSathyanarayana** [reply]
@borisno2 - Is your repository compatible with the latest version of Keystone-6?

**Reply by borisno2** [reply]
Hi @VinayaSathyanarayana, it certainly is, I updated it on the weekend...

**Comment by VinayaSathyanarayana** [comment]
Thanks @borisno2

**Comment by xmarkclx** [comment]
Along with supporting Next-Auth on admin panel, I would like to vote to have Next-Auth support for the client API (say NextJS)

**Reply by borisno2** [reply]
Hey @xmarkclx the package mentioned above (https://github.com/OpenSaasAU/keystone-nextjs-auth) does both admin and client (using nextJS), and contains an example of both. Shout out if you need a hand, or if you have any questions.

**Reply by xmarkclx** [reply]
Hi @borisno2 , basically we have a seperate NextJS app (in Vercel) and a Keystone 6 app (in Azure).

Do you mean to say that I can use this package to add social auth to the NextJS app (in Vercel)?

You know what, I am actually willing to pay you to integrate it to our demo app if you are available. :)

**Reply by borisno2** [reply]
Yeah it can add social auth to both the keystone admin ui and the nextjs app... if you are in the keystone slack feel free to reach out there... 

Or reach out through my website https://www.opensaas.com.au

happy to help out.
